### PR TITLE
rustdoc-json: change item ID's repr from a string to an int

### PR DESCRIPTION
**Pull request by its-the-shrimp**
Following [this discussion on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/266220-t-rustdoc/topic/Optimizing.20the.20.60Id.60.20type.20in.20.60rustdoc-types.60), I've changed the repr of `rustdoc_json_types::Id` from a String to a u32, by adding a `clean::ItemId` interner to `JsonRenderer`

r? @aDotInTheVoid

### CODE REVIEW

File: src/librustdoc/clean/types.rs [hunk 1]
```diff
@@ -57,20 +57,34 @@ pub(crate) type ItemIdSet = FxHashSet<ItemId>;
 #[derive(Debug, Clone, PartialEq, Eq, Hash, Copy)]
 pub(crate) enum ItemId {
     /// A "normal" item that uses a [`DefId`] for identification.
-    DefId(DefId),
+    DefId(DefId, Option<Namespace>),
```
Comment (obi1kenobi) [inline src/librustdoc/clean/types.rs:60]: I'm confused by the ability to set a `None` namespace here. Could you elaborate on the semantics of the new value here?

Also, I believe some items can cause names to appear in multiple namespaces — for example, unit structs. What would the namespace value be set to in such a case?

File: src/rustdoc-json-types/lib.rs [hunk 2]
```diff
@@ -296,9 +296,9 @@ pub enum TypeBindingKind {
 /// Rustdoc makes no guarantees about the inner value of Id's. Applications
 /// should treat them as opaque keys to lookup items, and avoid attempting
 /// to parse them, or otherwise depend on any implementation details.
-#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
 // FIXME(aDotInTheVoid): Consider making this non-public in rustdoc-types.
-pub struct Id(pub String);
+pub struct Id(pub u32);
```
Comment (obi1kenobi) [inline src/rustdoc-json-types/lib.rs:301]: Is there a way to recover the string ID from the numeric ID given a rustdoc JSON file? Or are you confident that isn't a capability that would be useful in any use case, so it isn't necessary to add?

**Review by obi1kenobi — COMMENTED** [review]
I'm broadly speaking quite unfamiliar with the librustdoc portion of the codebase, so apologies for my ignorant questions here.

File: src/rustdoc-json-types/lib.rs (hunk 2 above)
Comment (aDotInTheVoid) [reply src/rustdoc-json-types/lib.rs:301]: What do you mean "recover the string ID"? If you want to find what the Id would have been under the old schem, then no, this won't be possible, but I'm fine with that.

> [Rustdoc makes no guarantees about the inner value of Id’s. Applications should treat them as opaque keys to lookup items, and avoid attempting to parse them, or otherwise depend on any implementation details.](https://docs.rs/rustdoc-types/0.29.1/rustdoc_types/struct.Id.html)

File: src/librustdoc/formats/cache.rs [hunk 3]
```diff
@@ -27,8 +28,9 @@ use crate::visit_lib::RustdocEffectiveVisibilities;
 /// to be a fairly large and expensive structure to clone. Instead this adheres
 /// to `Send` so it may be stored in an `Arc` instance and shared among the various
 /// rendering threads.
-#[derive(Default)]
-pub(crate) struct Cache {
+pub(crate) struct Cache<'tcx> {
+    pub(crate) tcx: TyCtxt<'tcx>,
```
Comment (aDotInTheVoid) [inline src/librustdoc/formats/cache.rs:32]: Changing this is super-invasive. While it's possible that `Cache` should hold a `TyCtxt`, that should be changed in it's own PR (or at least commit), as it has implications for all of rustdoc, and should be discussed with the other maintainers. Doing it in the same commit makes this super-tricky to review.

File: src/librustdoc/formats/cache.rs [hunk 4]
```diff
@@ -77,6 +79,9 @@ pub(crate) struct Cache {
     /// Cache of where documentation for primitives can be found.
     pub(crate) primitive_locations: FxHashMap<clean::PrimitiveType, DefId>,
 
+    /// Used for [`Cache::intern_item_id`] to create [`rustdoc_json_types::Id`]s
+    item_id_interner: Lock<FxIndexSet<ItemId>>,
```
Comment (aDotInTheVoid) [inline src/librustdoc/formats/cache.rs:83]: Why are we using an `IndexSet`, instead of a `HashMap<ItemId, rustdoc_json_types::Id>` here? The latter seems far clearer (in my mind, at least).

File: src/librustdoc/json/conversions.rs [hunk 5]
```diff
@@ -105,37 +102,37 @@ impl JsonRenderer<'_> {
...
+                path: self.cache.tcx.def_path(did).to_string_no_crate_verbose(),
             },
         }
     }
 }
 
-pub(crate) trait FromWithTcx<T> {
-    fn from_tcx(f: T, tcx: TyCtxt<'_>) -> Self;
+pub(crate) trait FromWithCache<T> {
+    fn from_with_cache(f: T, cache: &Cache<'_>) -> Self;
```
Comment (aDotInTheVoid) [inline src/librustdoc/json/conversions.rs:113]: I don't think this is the right approach to threading some mutable state through the clean->JSON pass. For one, it hides the fact that `Cache` is mutable, why doesn't it take `&mut Cache`, so we can remove the `Lock` on the ItemId interning map.

But more broadly, I don't think we should be adding more things to `Cache` that are only needed here. I think a better aproach would be to do

```rust
// TODO: Better name for this
struct JsonItemConverter<'tcx> {
   tcx: TyCtxt<'tcx>,
   cache: Cache,
   id_interner: FxHashMap<ItemId, rustdoc_json_types::Id>
}
```

and then pass that to all the converting methods (either via a trait, or via methods on `impl JsonItemConverter`

File: src/librustdoc/clean/types.rs (hunk 1 above)
Comment (aDotInTheVoid) [inline src/librustdoc/clean/types.rs:60]: Why is is necessary to change this? And if it is, could you please do that is a separate PR, as that would make it much easier to review.

**Review by aDotInTheVoid — COMMENTED** [review]
This PR's doing alot, and needs major rework, so I've only tried to review it at a high level, as I don't think the in-the-weeds stuff is worth it yet.

1. The changes to `ItemId` are intrusive to most of the rustdoc codebase, and not motivated at all. If these are nessessary, could that be explained, and send in a different, tighter-scoped, PR (that would be landed before this)
2. Making `Cache` be the only source of state for the `clean`->`json` lowering is quite a bad idea:
    - It requires adding a `TyCtxt`, which produces more changes to the rest of rustdoc.
    - It adds the interner field which isn't used for HTML, but is still present, which is unfortunate.
    - In present, it requires adding a `Lock`, but I don't think this is inherent.
3. There seem to be unrelated cleanups (like the removal of `FormatRenderer::cache`). These are should be merged into rustdoc, but not here! Again, the answer is a smaller, seperate PR, that just does one thing, and will be much easier to review.

File: src/librustdoc/clean/types.rs (hunk 1 above)
Comment (its-the-shrimp) [reply src/librustdoc/clean/types.rs:60]: Without this disambiguation, rustdoc would start confusing things like a primitive & a module that documents its, a struct & a function with the same name, etc., it looked to me like a case of DefIds not carrying namespace info

File: src/librustdoc/formats/cache.rs (hunk 4 above)
Comment (its-the-shrimp) [reply src/librustdoc/formats/cache.rs:83]: IndexSet returns the index of the item that's been added, with a HashMap it'd need an additional counter attached to it that'd be incremented only if the Id hasn't been added to the map yet

Comment (aDotInTheVoid) [reply src/librustdoc/formats/cache.rs:83]: You can just use the `.len()` for this (which is what [`IndexMap`](https://github.com/indexmap-rs/indexmap/blob/abdd155f73331040991f24eaff6f37976097711e/src/map/core/raw.rs#L70) does internally). Yes this means we need to implement a bit more logic in rustdoc, but it make the nature as a Map clearer (rather than being a set), and means you don't have to pay the overhead of `IndexMap` (vs hashbrown's HashMap)

File: src/librustdoc/clean/types.rs (hunk 1 above)
Comment (its-the-shrimp) [reply src/librustdoc/clean/types.rs]: Initial test failures led me to believe that DefIds didn't carry namespace information, but that guess turned out incorrect, I've changed the impl now

Comment (its-the-shrimp) [reply src/librustdoc/clean/types.rs]: nevermind, I've realised later that the issues that led me to believe that DefIds don't carry namespace info were related to imports of different items with the same name, I've changed the impl now

**Comment by jalil-salame** [comment]
Some tests are failing to compile due to them expecting the id to be `Id(String)`: https://github.com/rust-lang/rust/actions/runs/10769404806/job/29860558315?pr=130078#step:25:6101

**Comment by its-the-shrimp** [comment]
@rustbot ready

This time the interner is restricted to the JsonRenderer, however I still couldn't get rid of the lock on the interner because of borrow checking errors

**Comment by its-the-shrimp** [comment]
It could be possible to get rid of it if we were to change the way `formats::run_format` drives the renderer forward, currently it clones the renderer at least once (and for JSON it'll do so only once), which means the ID interner has to be in an Rc, which is unfortunate since that clone is completely redundant

**Comment by its-the-shrimp** [comment]
Or I can do it in this PR, the change should be very minimal

**Comment by its-the-shrimp** [comment]
should the FORMAT_VERSION be bumped for this?

**Comment by jalil-salame** [comment]
> should the FORMAT_VERSION be bumped for this?

Most likely yes, it is an API change (`rustdoc-types` will be unable to parse files with `Id(u32)`).

**Comment by aDotInTheVoid** [comment]
Yeah, bumping the format version *needs* to happen whenever we do something that breaks deserialisation (and I don’t know any times it’s been done for other reasons, but would be interested).

> The representation of IDs doesn't have any guarantees associated with it though

Previously we guaranteed they were strings.

**Review by aDotInTheVoid — APPROVED** [review]

**Comment by aDotInTheVoid** [comment]
@bors r+

Thanks for working on this, and sorry for the delay.
